bundles/org.eclipse.equinox.p2.director.app/src/org/eclipse/equinox/internal/p2/director/app/DirectorApplication.java

```diff
@@ -1058,6 +1063,13 @@ private void performProvisioningActions() throws CoreException {
 			context.setProperty(ProvisioningContext.FOLLOW_REPOSITORY_REFERENCES, String.valueOf(followReferences));
 			context.setProperty(FOLLOW_ARTIFACT_REPOSITORY_REFERENCES, String.valueOf(followReferences));
 
+			if (addJreIU) {
+				List<IInstallableUnit> extraUnits = new ArrayList<>();
+				IInstallableUnit jreiU = JREAction.createJreIU();
+				extraUnits.add(jreiU);
+				context.setExtraInstallableUnits(extraUnits);
```

Comment:
I have to look in more detail, but hopefully there are no other places that call setExtraInstallableUnits such that this might be replacing those or vice versa.

Reply:
Here jreiU should be jreIU and I assume you will  renamed the method to createJREIU.

Reply:
I can see others also setting `extraIUs`.
I think it is a problem.
As I can see code for 	
```
	public void setExtraInstallableUnits(List<IInstallableUnit> extraIUs) {
		this.extraIUs.clear();
		//copy the list to prevent future client tampering
		if (extraIUs != null)
			this.extraIUs.addAll(extraIUs);
	}
```
If I can remove `this.extraIUs.clear();`  others `extraUnits` will be added with it. But I don't know the reason why we need to clear here before adding `extraIUs`

Reply:
I think it's okay how it is.  I see these calls:

![image](https://github.com/eclipse-equinox/p2/assets/208716/dbc4af2e-e1fc-429d-844e-472da37935a8)

In all these places the new context has just been created.

We should definitely **not** change the behavior of org.eclipse.equinox.p2.engine.ProvisioningContext.setExtraInstallableUnits(List<IInstallableUnit>) we cannot know what other clients do.  Looking at the implementation of org.eclipse.equinox.p2.engine.ProvisioningContext.getExtraInstallableUnits() I see it actually returns a modifiable list so one could easily do an add, but as I said, in both these internal uses the context has just been created.

Reply:
Yes. I just debugged the flow. There is no problem.